### PR TITLE
Gchehade/streamable state

### PR DESCRIPTION
**Pull request by poolpOrg**

### CODE REVIEW

File: repository/state/state.go [hunk 1]
```diff
@@ -141,37 +129,256 @@ func (st *State) getOrCreateIdForChecksum(checksum objects.Checksum) uint64 {
...
+	}
+	for _, checksum := range st.Metadata.Extends {
+		if _, err := w.Write(checksum[:]); err != nil {
+			return fmt.Errorf("failed to write checksum: %w", err)
+		}
+	}
+
+	if err := serializeMapping(w, st.DeletedSnapshots, func(key uint64) error {
+		return writeUint64(key)
+	}, func(value time.Time) error {
```
Comment (omar-polo) [inline repository/state/state.go:185]: nit: could pass writeUint64 directly instead of wrapping it.

File: repository/state/state.go [hunk 2]
```diff
@@ -141,37 +129,256 @@ func (st *State) getOrCreateIdForChecksum(checksum objects.Checksum) uint64 {
...
+		{"Snapshots", st.Snapshots},
+		{"Signatures", st.Signatures},
+		{"Errors", st.Errors},
 	}
 
-	st.rebuildChecksums()
+	for _, m := range mappings {
+		if err := serializeMapping(w, m.data, writeUint64, func(v Location) error {
+			return writeLocation(v)
+		}); err != nil {
```
Comment (omar-polo) [inline repository/state/state.go:214]: nit: could pass `writeLocation` directly instead of wrapping it in a no-op.

**Review by omar-polo — APPROVED** [review]
two minor nitpicks; but looks good to me.  Love the change, looking forward to what it'll unlock :)

File: repository/state/state.go (hunk 1 above)
Comment (poolpOrg) [reply repository/state/state.go]: fixed

File: repository/state/state.go (hunk 2 above)
Comment (poolpOrg) [reply repository/state/state.go]: fixed !

**Pull request merged**